### PR TITLE
Store user info in DB and make them searchable in ACL UI

### LINKED ISSUE #951: Implement some way to add users in the ACL selector
Currently we have a bunch of dummy users you can select. That obviously has to be replaced, but with what? The main problem, as described in #671 is that we don't just have a list of all users. And we can't require our login system to provide that. Further, it is even questionable from a privacy standpoint whether that's ok (though for the ETH it would be fine).

A few ideas:

- Tobira can just remember all users that ever logged into Tobira and make that searchable in the UI.
- We can have an import command, e.g. `tobira import-known-users` or something like that where organizations can ingest such a list. That's mostly useful for the beginning period where very few users already logged into Tobira. But it could also be used in a cronjob to push up to date user info.

Regarding privacy one could make it so that users have to enter an exact username instead.

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Fixes #951 (also removing our dummy users from the code, unblocking the release)

The main goal of this PR was to make the ACL UI user selection work nicely. I.e. that one can search through users to add user entries to the ACL. Previously we had a bunch of hardcoded dummy users (which also made us unable to release Tobira). This is now done, but lots of related changed had to be done as well. There is also a configuration option which controls whether users can actually be searched by name or whether they can only be found by typing the exact username/email. The latter mode is the default for data privacy reasons.

User information is remembered whenever a user logs into Tobira or does anything there. It is also possible to import users from a JSON file. 

**Important**: this PR introduces a breaking change as it makes a "user role" mandatory for users. See second commit. I doubt this is a problem for anyone, but it's still technically breaking.

---

This can probably be reviewed commit by commit. The commit messages should be read for sure. However, the changes to `ui/Access.tsx` are likely very annoying to review because it's also lots of refactoring, over multiple commits. So yeah, not 100% sure how to best approach that.

---

Finally, there are a few things that still have to be improved. But not in this PR, it's already large enough. I will create issues for these after merging.

- [ ] Potentially add some basic user stats to `/~metrics`, e.g. "active user in last 24h
- [ ] Re-add the paste functionality
- [ ] Stop sending a list of all known groups to the frontend, that makes stuff slow.

### CODE REVIEW

**Comment by owi92** [comment]
Only testing deployment with dummy users in a first baby step. So I have to log in with each user before I can find them, so far so good. However, I find the default for searching for users somewhat unintuitive, where typing `sabine` will recommend `Sabine Rudolfs`, but typing `sabine ` (with a space) does not. Neither does `sabine rudolfs`. Typing `jose` finds José, but typing `josé` does not. `bjork` doesn't find anything, but `Björk` does.
I understand that this is by design, but it is a little confusing at first. Is this really the best practice or required by any institution? I'm just imagining a lecturer/tutor asking their students to each give them their "tobira username" or sth along those lines, which seems really impractical to me. I suppose since they can also be found by their email address, it's not such a big deal.
I'm sure you thought about this a bunch, so if this is really the way it's supposed to be, then I'm not complaining.

**Comment by LukasKalbertodt** [comment]
> typing `sabine` will recommend `Sabine Rudolfs`, but typing `sabine ` (with a space) does not. 

Oh ok yeah, I suppose we can trim whitespace for sure.

> Neither does `sabine rudolfs`. Typing `jose` finds José, but typing `josé` does not. `bjork` doesn't find anything, but `Björk` does. 

But yeah all of that has to do with the fact that users can only be found by their exact username or email (well, case insensitive). I hope the message in the searchable select make that clear enough? I think part of the confusion here stems from the fact that our dummy users have very "nice" usernames. In practice, almost all users will have something a bit more ugly like `lkalbertodt` or event something terrible with a big number or so. Further confusion is added because our dummy user Björk actually has the username `björk` (with umlaut!) which is likely very uncommon in practice. Since our dummy usernames are so close to the display names, I suspect it's easy to assume simple variations would also find the uesrs. With more ugly usernames I think people won't assume that anymore.

The "exact username or email" is the default behavior as it's the more safe one. And for what it's worth: our test deployment will use the "free text search", this simply isn't active on this PR's test deployment yet due to our particular deployment specifics.

I hope that alleviates some of your worries?

**Review by owi92 — APPROVED** [review]
Phew that was a chucky review^^ But you did a very good job on this (unsurprisingly ;). I'm glad I only had to review this, and not implement it myself! Everything seems to be working, and I only have very few comments. I think none of which are mandatory changes, so I'm already approving. Though as I said, I might do some more testing as well.

File: backend/src/auth/mod.rs [hunk 1]
```diff
@@ -157,6 +157,38 @@ impl AuthMode {
...
+        let Some(user_role) = roles.by_ref().find(is_user_role) else {
+            warn!("User '{username}' has no user role, but it needs exactly one. {note}");
+            return None;
+        };
+
+
+        if let Some(extra) = roles.find(is_user_role) {
+            warn!(
+                "User '{username}' has multiple user roles ({user_role} and {extra}) \
+                    but there should be one unique user role per user. {note}",
```
[the remark this answers is not in the document]
Comment (owi92) [reply backend/src/auth/mod.rs:186]: Btw of course I'm not suggesting to drop the `{note}` part!

File: backend/src/api/model/acl.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    //         map.entry(role).or_default().push(action.into());
+    //     }
+    // }
+
+    // Ok(map.into_iter().map(|(role, actions)| AclItem {
+    //     role: role.into(),
+    //     // Roles are unique so we can `remove` her to avoid cloning.
+    //     label: labels.remove(role).flatten(),
+    //     actions,
+    // }).collect())
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply backend/src/api/model/acl.rs:117]: Woops!

File: frontend/src/ui/Access.tsx [hunk 3]
```diff
@@ -68,27 +78,34 @@ type AclSelectorProps = {
...
-            readRoles: new Set(acl.readRoles),
-            writeRoles: new Set(acl.writeRoles),
-        };
+        const copy = new Map([...acl.entries()].map(([role, value]) => [role, {
+            actions: new Set(value.actions),
+            info: value.info == null ? null : {
+                label: value.info.label,
+                implies: value.info.implies,
+                large: value.info.large,
+            },
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/ui/Access.tsx:97]: Oh yes :D

File: frontend/src/i18n/locales/de.yaml [hunk 4]
```diff
@@ -291,6 +291,11 @@ manage:
   access:
     authorized-groups: Autorisierte Gruppen
     authorized-users: Autorisierte Personen
+    users-no-options:
+      initial-searchable: Nach Name suchen oder exakten Nutzernamen/exakte E-Mail angeben
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/i18n/locales/de.yaml:295]: I got really confused thinking about this, so I asked my German-teacher mom :grimacing: So yes, if one would write this as a whole sentence, "Sie sollen nach einem/dem Namen suchen" then "Namen" is the correct form. But Nominativ is fine in this case too she says. And my intuition also (clearly) told me "Name". So I would keep it.

**Pull request merged**